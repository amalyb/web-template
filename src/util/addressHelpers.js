/**
 * Address normalization and mapping utilities
 */

// Normalize address data by trimming spaces and converting to standard formats
export const normalizeAddress = (address) => {
  if (!address) return null;

  const normalized = {
    fullName: address.fullName?.trim().replace(/\s+/g, ' ') || '',
    country: address.country?.toUpperCase() || 'US',
    street1: address.street1?.trim().replace(/\s+/g, ' ') || '',
    street2: address.street2?.trim().replace(/\s+/g, ' ') || undefined, // Always send undefined when missing
    city: address.city?.trim().replace(/\s+/g, ' ') || '',
    state: address.state?.trim().toUpperCase() || '',
    zip: address.zip?.trim().replace(/\s+/g, '') || '',
    email: address.email?.trim().toLowerCase() || '',
    phone: normalizePhone(address.phone) || '',
  };

  // Clean up empty strings to undefined for optional fields
  if (!normalized.street2) normalized.street2 = undefined;
  if (!normalized.phone) normalized.phone = undefined;

  return normalized;
};

// Normalize phone number to E.164 format
export const normalizePhone = (phone) => {
  if (!phone) return '';
  
  // Remove extensions (x123, ext 123, etc.) and reject them
  let cleaned = phone.replace(/\s*(x|ext|extension)\s*\d+.*$/i, '');
  
  // Remove all non-digit characters except +
  cleaned = cleaned.replace(/[^\d+]/g, '');
  
  // If it doesn't start with +, assume US number and add +1
  if (!cleaned.startsWith('+')) {
    // Remove leading 1 if present
    if (cleaned.startsWith('1') && cleaned.length === 11) {
      cleaned = cleaned.substring(1);
    }
    cleaned = '+1' + cleaned;
  }
  
  // Validate E.164 format (1-15 digits after +)
  const e164Regex = /^\+[1-9]\d{1,14}$/;
  if (!e164Regex.test(cleaned)) {
    return phone; // Return original if normalization fails
  }
  
  return cleaned;
};

// Map normalized address to Stripe billing format
export const mapToStripeBilling = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    address: {
      line1: normalized.street1,
      line2: normalized.street2 || undefined, // Always send undefined when missing
      city: normalized.city,
      state: normalized.state,
      postal_code: normalized.zip,
      country: normalized.country,
    },
  };
};

// Map normalized address to Shippo/UPS/FedEx format
export const mapToShippo = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    street1: normalized.street1,
    street2: normalized.street2 || '',
    city: normalized.city,
    state: normalized.state,
    zip: normalized.zip,
    country: normalized.country,
  };
};

// Check if address is a PO Box (for courier routing)
export const isPOBox = (address) => {
  if (!address?.street1) return false;
  return /^(P(OST)?\.?\s*O(FFICE)?\.?\s*BOX)\b/i.test(address.street1);
};

// Map normalized address to UPS format
export const mapToUPS = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    address_line1: normalized.street1,
    address_line2: normalized.street2 || '',
    city: normalized.city,
    state_province_code: normalized.state,
    postal_code: normalized.zip,
    country_code: normalized.country,
  };
};

// Map normalized address to FedEx format
export const mapToFedEx = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    contact: {
      person_name: normalized.fullName,
      email_address: normalized.email,
      phone_number: normalized.phone,
    },
    address: {
      street_lines: [normalized.street1, normalized.street2].filter(Boolean),
      city: normalized.city,
      state_or_province_code: normalized.state,
      postal_code: normalized.zip,
      country_code: normalized.country,
    },
  };
};

// Validate address completeness
export const validateAddress = (address, requiredFields = {}) => {
  const defaults = {
    fullName: true,
    country: true,
    street1: true,
    city: true,
    state: true,
    zip: true,
    email: true,
    phone: false,
  };
  
  const requirements = { ...defaults, ...requiredFields };
  const errors = {};
  
  if (requirements.fullName && !address.fullName?.trim()) {
    errors.fullName = 'Full name is required';
  }
  
  if (requirements.country && !address.country?.trim()) {
    errors.country = 'Country is required';
  }
  
  if (requirements.street1 && !address.street1?.trim()) {
    errors.street1 = 'Street address is required';
  }
  
  if (requirements.city && !address.city?.trim()) {
    errors.city = 'City is required';
  }
  
  if (requirements.state && !address.state?.trim()) {
    errors.state = 'State/Region is required';
  }
  
  if (requirements.zip && !address.zip?.trim()) {
    errors.zip = 'ZIP/Postal code is required';
  }
  
  if (requirements.email && !address.email?.trim()) {
    errors.email = 'Email is required';
  } else if (address.email && !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(address.email)) {
    errors.email = 'Please enter a valid email address';
  }
  
  if (requirements.phone && !address.phone?.trim()) {
    errors.phone = 'Phone number is required';
  }
  
  return {
    isValid: Object.keys(errors).length === 0,
    errors,
  };
};

// Copy billing address to shipping address
export const copyBillingToShipping = (billingAddress) => {
  const normalized = normalizeAddress(billingAddress);
  if (!normalized) return null;

  return {
    shipping: {
      fullName: normalized.fullName,
      country: normalized.country,
      street1: normalized.street1,
      street2: normalized.street2,
      city: normalized.city,
      state: normalized.state,
      zip: normalized.zip,
      email: normalized.email,
      phone: normalized.phone,
    },
  };
};
