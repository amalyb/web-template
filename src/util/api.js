// These helpers are calling this template's own server-side routes
// so, they are not directly calling Marketplace API or Integration API.
// You can find these api endpoints from 'server/api/...' directory

import axios from 'axios';
import appSettings from '../config/settings';
import { types as sdkTypes, transit } from './sdkLoader';
import Decimal from 'decimal.js';

export const apiBaseUrl = marketplaceRootURL => {
  const port = process.env.REACT_APP_DEV_API_SERVER_PORT;
  const useDevApiServer = process.env.NODE_ENV === 'development' && !!port;

  // In development, the dev API server is running in a different port
  if (useDevApiServer) {
    return `http://localhost:${port}`;
  }

  // Otherwise, use the given marketplaceRootURL parameter or the same domain and port as the frontend
  return marketplaceRootURL ? marketplaceRootURL.replace(/\/$/, '') : `${window.location.origin}`;
};

// Axios client for API calls with Basic authentication
const apiBase = process.env.REACT_APP_API_BASE_URL || '/api';
export const apiClient = axios.create({ baseURL: apiBase });

apiClient.interceptors.request.use((config) => {
  const u = process.env.REACT_APP_BASIC_AUTH_USERNAME;
  const p = process.env.REACT_APP_BASIC_AUTH_PASSWORD;
  if (u && p) {
    const token = btoa(`${u}:${p}`);
    config.headers = { ...config.headers, Authorization: `Basic ${token}` };
  }
  return config;
});

// Application type handlers for JS SDK.
//
// NOTE: keep in sync with `typeHandlers` in `server/api-util/sdk.js`
export const typeHandlers = [
  // Use Decimal type instead of SDK's BigDecimal.
  {
    type: sdkTypes.BigDecimal,
    customType: Decimal,
    writer: v => new sdkTypes.BigDecimal(v.toString()),
    reader: v => new Decimal(v.value),
  },
];

const serialize = data => {
  return transit.write(data, { typeHandlers, verbose: appSettings.sdk.transitVerbose });
};

const deserialize = str => {
  return transit.read(str, { typeHandlers });
};

const methods = {
  POST: 'POST',
  GET: 'GET',
  PUT: 'PUT',
  PATCH: 'PATCH',
  DELETE: 'DELETE',
};

// If server/api returns data from SDK, you should set Content-Type to 'application/transit+json'
const request = (path, options = {}) => {
  const url = `${apiBaseUrl()}${path}`;
  const { credentials, headers, body, ...rest } = options;

  // If headers are not set, we assume that the body should be serialized as transit format.
  const shouldSerializeBody =
    (!headers || headers['Content-Type'] === 'application/transit+json') && body;
  const bodyMaybe = shouldSerializeBody ? { body: serialize(body) } : {};

  const fetchOptions = {
    credentials: credentials || 'include',
    // Since server/api mostly talks to Marketplace API using SDK,
    // we default to 'application/transit+json' as content type (as SDK uses transit).
    headers: headers || { 'Content-Type': 'application/transit+json' },
    ...bodyMaybe,
    ...rest,
  };

  return window.fetch(url, fetchOptions).then(res => {
    const contentTypeHeader = res.headers.get('Content-Type');
    const contentType = contentTypeHeader ? contentTypeHeader.split(';')[0] : null;

    if (res.status >= 400) {
      return res.json().then(data => {
        let e = new Error();
        e = Object.assign(e, data);

        throw e;
      });
    }
    if (contentType === 'application/transit+json') {
      return res.text().then(deserialize);
    } else if (contentType === 'application/json') {
      return res.json();
    }
    return res.text();
  });
};

// Keep the previous parameter order for the post method.
// For now, only POST has own specific function, but you can create more or use request directly.
export const post = (path, body, options = {}) => {
  const requestOptions = {
    ...options,
    method: methods.POST,
    body,
  };

  return request(path, requestOptions);
};

// Fetch transaction line items from the local API endpoint.
//
// See `server/api/transaction-line-items.js` to see what data should
// be sent in the body.
export const transactionLineItems = (payload) =>
  apiClient.post('/transaction-line-items', payload);

// Initiate a privileged transaction.
//
// With privileged transitions, the transactions need to be created
// from the backend. This endpoint enables sending the order data to
// the local backend, and passing that to the Marketplace API.
//
// See `server/api/initiate-privileged.js` to see what data should be
// sent in the body.
export const initiatePrivileged = body => {
  return post('/api/initiate-privileged', body);
};

// Transition a transaction with a privileged transition.
//
// This is similar to the `initiatePrivileged` above. It will use the
// backend for the transition. The backend endpoint will add the
// payment line items to the transition params.
//
// See `server/api/transition-privileged.js` to see what data should
// be sent in the body.
export const transitionPrivileged = body => {
  return post('/api/transition-privileged', body);
};

// Create user with identity provider (e.g. Facebook or Google)
//
// If loginWithIdp api call fails and user can't authenticate to Marketplace API with idp
// we will show option to create a new user with idp.
// For that user needs to confirm data fetched from the idp.
// After the confirmation, this endpoint is called to create a new user with confirmed data.
//
// See `server/api/auth/createUserWithIdp.js` to see what data should
// be sent in the body.
export const createUserWithIdp = body => {
  return post('/api/auth/create-user-with-idp', body);
};

// Ensure phone number is saved to protectedData
// This can be called after sign-up or profile updates to guarantee the phone number is stored correctly
export const ensurePhoneNumber = phoneNumber => {
  if (!phoneNumber) {
    console.warn('‚ö†Ô∏è [api] No phone number provided to ensurePhoneNumber');
    return Promise.resolve();
  }
  
  console.log('üì± [api] Ensuring phone number is saved to protectedData:', phoneNumber);
  return post('/api/ensure-phone-number', { phoneNumber })
    .then(() => {
      console.log('‚úÖ [api] Phone number ensured in protectedData');
    })
    .catch(error => {
      console.warn('‚ö†Ô∏è [api] Failed to ensure phone number:', error.message);
      throw error;
    });
};
