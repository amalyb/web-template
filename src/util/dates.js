import moment from 'moment-timezone/builds/moment-timezone-with-data-10-year-range.min';

/**
 * Input names for the DateRangePicker from react-dates.
 */
export const START_DATE = 'startDate';
export const END_DATE = 'endDate';

/**
 * Check that the given parameter is a Date object.
 *
 * @param {Date} object that should be a Date.
 *
 * @returns {boolean} true if given parameter is a Date object.
 */
export const isDate = d =>
  d && Object.prototype.toString.call(d) === '[object Date]' && !Number.isNaN(d.getTime());

/**
 * Check if the given parameters represent the same Date value (timestamps are compared)
 *
 * @param {Date} first param that should be a Date and it should have same timestamp as second param.
 * @param {Date} second param that should be a Date and it should have same timestamp as second param.
 *
 * @returns {boolean} true if given parameters have the same timestamp.
 */
export const isSameDate = (a, b) => a && isDate(a) && b && isDate(b) && a.getTime() === b.getTime();

/**
 * Compare is dateA is after dateB
 *
 * @param {Date} dateA date instance
 * @param {Date} dateB date instance
 *
 * @returns {Date} true if dateA is after dateB
 */
export const isAfterDate = (dateA, dateB) => {
  return moment(storedAt).isAfter(moment(dateB));
};

////////////////////////////////////////////////////////////////////
// Manipulate time: time-of-day between different time zones etc. //
////////////////////////////////////////////////////////////////////

/**
 * Returns a new date, which indicates the same time of day in a given time zone
 * as given date is in local time zone
 *
 * @param {Date} date
 * @param {String} timeZone
 *
 * @returns {Date} date in given time zone
 */
export const timeOfDayFromLocalToTimeZone = (date, timeZone) => {
  return moment.tz(moment(date).format('YYYY-MM-DD HH:mm:ss'), timeZone).toDate();
};

/**
 * Returns a new date, which indicates the same time of day in a local time zone
 * as given date is in specified time zone
 *
 * @param {Date} date
 * @param {String} timeZone
 *
 * @returns {Date} date in given time zone
 */
export const timeOfDayFromTimeZoneToLocal = (date, timeZone) => {
  return moment(
    moment(date)
      .tz(timeZone)
      .format('YYYY-MM-DD HH:mm:ss')
  ).toDate();
};

/**
 * Get start of time unit (e.g. start of day)
 *
 * @param {Date} date date instance to be converted
 * @param {String} unit time-unit (e.g. "day")
 * @param {String} timeZone time zone id
 *
 * @returns {Date} date object converted to the start of given unit
 */
export const getStartOf = (date, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();

  return m.startOf(unit).toDate();
};

/**
 * Adds time-units to the date
 *
 * @param {Date} date date to be manipulated
 * @param {int} offset offset of time-units (e.g. "3" days)
 * @param {String} unit time-unit (e.g. "days")
 * @param {String} timeZone time zone name
 *
 * @returns {Date} date with given offset added
 */
export const addTime = (date, offset, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();
  return m.add(offset, unit).toDate();
};

/**
 * Subtract time-units from the date
 *
 * @param {Date} date date to be manipulated
 * @param {int} offset offset of time-units (e.g. "3" days)
 * @param {String} unit time-unit (e.g. "days")
 * @param {String} timeZone time zone name
 *
 * @returns {Date} date with given offset subtracted
 */
export const subtractTime = (date, offset, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();
  return m.subtract(offset, unit).toDate();
};

///////////////
// Durations //
///////////////

/**
 * Calculate the number of days between the given dates
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period. NOTE: with daily
 * bookings, it is expected that this date is the exclusive end date,
 * i.e. the last day of the booking is the previous date of this end
 * date.
 *
 * @throws Will throw if the end date is before the start date
 * @returns {Number} number of days between the given dates
 */
export const daysBetween = (startDate, endDate) => {
  const days = moment(endDate).diff(startDate, 'days');
  if (days < 0) {
    throw new Error('End date cannot be before start date');
  }
  return days;
};

/**
 * Calculate the number of minutes between the given dates
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period.
 *
 * @throws Will throw if the end date is before the start date
 * @returns {Number} number of minutes between the given Date objects
 */
export const minutesBetween = (startDate, endDate) => {
  const minutes = moment(endDate).diff(startDate, 'minutes');
  if (minutes < 0) {
    throw new Error('End Date cannot be before start Date');
  }
  return minutes;
};

/**
 * Calculate the difference between the given dates
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period.
 *
 * @returns {Number} time difference between the given Date objects using given unit
 */
export const diffInTime = (startDate, endDate, unit, useFloat = false) => {
  return startDate.diff(endDate, unit, useFloat);
};

////////////////////////////
// Parsing and formatting //
////////////////////////////

/**
 * Format the given date
 *
 * @param {Object} intl Intl object from react-intl
 * @param {String} todayString translation for the current day
 * @param {Date} d Date to be formatted
 *
 * @returns {String} formatted date
 */
export const formatDate = (intl, todayString, d) => {
  const paramsValid = intl && d instanceof Date && typeof todayString === 'string';
  if (!paramsValid) {
    throw new Error(`Invalid params for formatDate: (${intl}, ${todayString}, ${d})`);
  }

  // By default we can use moment() directly but in tests we need to use a specific dates.
  // fakeIntl used in tests contains now() function wich returns predefined date
  const now = intl.now ? moment(intl.now()) : moment();
  const formattedTime = intl.formatTime(d);
  let formattedDate;

  if (now.isSame(d, 'day')) {
    // e.g. "Today, 9:10pm"
    formattedDate = todayString;
  } else if (now.isSame(d, 'week')) {
    // e.g. "Wed, 8:00pm"
    formattedDate = intl.formatDate(d, {
      weekday: 'short',
    });
  } else if (now.isSame(d, 'year')) {
    // e.g. "Aug 22, 7:40pm"
    formattedDate = intl.formatDate(d, {
      month: 'short',
      day: 'numeric',
    });
  } else {
    // e.g. "Jul 17 2016, 6:02pm"
    const date = intl.formatDate(d, {
      month: 'short',
      day: 'numeric',
    });
    const year = intl.formatDate(d, {
      year: 'numeric',
    });
    formattedDate = `${date} ${year}`;
  }

  return `${formattedDate}, ${formattedTime}`;
};

/**
 * Converts string given in ISO8601 format to date object.
 * This is used e.g. when when dates are parsed form urlParams
 *
 * @param {String} dateString in 'YYYY-MM-DD'format
 *
 * @returns {Date} parsed date object
 */
export const parseDateFromISO8601 = dateString => {
  return moment(dateString, 'YYYY-MM-DD').toDate();
};

/**
 * Converts date to string ISO8601 format ('YYYY-MM-DD').
 * This string is used e.g. in urlParam.
 *
 * @param {Date} date
 *
 * @returns {String} string in 'YYYY-MM-DD'format
 */
export const stringifyDateToISO8601 = date => {
  return moment(date).format('YYYY-MM-DD');
};

/**
 * Formats string ('YYYY-MM-DD') to UTC format ('0000-00-00T00:00:00.000Z').
 * This is used in search query.
 *
 * @param {String} string in 'YYYY-MM-DD'format
 *
 * @returns {String} string in '0000-00-00T00:00:00.000Z' format
 */
export const formatDateStringToUTC = dateString => {
  return moment.utc(dateString).toDate();
};

/**
 * Formats date to into multiple different ways:
 * - date "Mar 24"
 * - time "8:07 PM"
 * - dateAndTime: "Mar 24, 8:07 PM"
 *
 * @param {Object} intl Intl object from react-intl
 * @param {Date} date to be formatted
 *
 * @returns {Object} "{ date, time, dateAndTime }"
 */
export const formatDateToText = (intl, date) => {
  return {
    date: intl.formatDate(date, {
      month: 'short',
      day: 'numeric',
    }),
    time: intl.formatDate(date, {
      hour: 'numeric',
      minute: 'numeric',
    }),
    dateAndTime: intl.formatTime(date, {
      month: 'short',
      day: 'numeric',
    }),
  };
};

//////////
// Misc //
//////////

/**
 * Format the given date to month id/string
 *
 * @param {Date} date to be formatted
 *
 * @returns {String} formatted month string
 */
export const monthIdString = date => moment(date).format('YYYY-MM');

/**
 * Format the given date to UTC month id/string
 *
 * @param {Date} date to be formatted
 *
 * @returns {String} formatted month string
 */
export const monthIdStringInUTC = date =>
  moment(date)
    .utc()
    .format('YYYY-MM');

/**
 * Formats string ('YYYY-MM-DD') to UTC format ('0000-00-00T00:00:00.000Z') and adds one day.
 * This is used as end date of the search query.
 * One day must be added because end of the availability is exclusive in API.
 *
 * @param {String} string in 'YYYY-MM-DD'format
 *
 * @returns {String} string in '0000-00-00T00:00:00.000Z' format
 */

export const getExclusiveEndDate = dateString => {
  return moment
    .utc(dateString)
    .add(1, 'days')
    .startOf('day')
    .toDate();
};
