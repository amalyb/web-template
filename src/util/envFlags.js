/**
 * Centralized environment flags for client-side code
 * Safely handles missing process object in browser environments
 */

export const IS_PROD = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'production');
export const IS_DEV  = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'development');
export const IS_TEST = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'test');
export const __DEV__ = !IS_PROD;
