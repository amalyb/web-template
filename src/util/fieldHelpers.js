import { SCHEMA_TYPE_MULTI_ENUM, SCHEMA_TYPE_TEXT } from './types';

const keyMapping = {
  userType: {
    wrapper: 'userTypeConfig',
    limitTo: 'limitToUserTypeIds',
    ids: 'userTypeIds',
  },
  listingType: {
    wrapper: 'listingTypeConfig',
    limitTo: 'limitToListingTypeIds',
    ids: 'listingTypeIds',
  },
};

const getEntityTypeRestrictions = (entityTypeKey, config) => {
  const { wrapper, limitTo, ids } = keyMapping[entityTypeKey];
  const isLimited = config && config[wrapper] && config[wrapper][limitTo];
  const limitToIds = config && config[wrapper] && config[wrapper][ids];

  return { isLimited, limitToIds };
};

/**
 * Returns the value of the attribute in extended data.
 * @param {*} data extended data containing the value
 * @param {*} key attribute key in extended data
 * @returns
 */
export const getFieldValue = (data, key) => {
  const value = data?.[key];
  return value != null ? value : null;
};

/**
 * Pick props for SectionMultiEnumMaybe and SectionTextMaybe display components.
 * @param {*} publicData entity public data containing the value(s) to be displayed
 * @param {*} metadata entity metadata containing the value(s) to be displayed
 * @param {*} fieldConfig custom field configuration object
 * @param {*} entityTypeKey the name of the key denoting the entity type in publicData.
 * Currently supports 'userType' and 'listingType'.
 * @returns an object with attributes 'schemaType', 'key', and 'heading', as well as either
 * - 'options' and 'selectedOptions' for SCHEMA_TYPE_MULTI_ENUM
 * - or 'text' for SCHEMA_TYPE_TEXT
 *
 */
export const pickCustomFieldProps = (publicData, metadata, fieldConfig, entityTypeKey) => {
  return fieldConfig?.reduce((pickedElements, config) => {
    const { key, enumOptions, schemaType, showConfig = {}, scope = 'public' } = config;
    const { displayInProfile } = showConfig;
    const { isLimited, limitToIds } = getEntityTypeRestrictions(entityTypeKey, config);
    const entityType = publicData && publicData[entityTypeKey];
    const isTargetEntityType = !isLimited || limitToIds.includes(entityType);

    const createFilterOptions = options =>
      options.map(o => ({ key: `${o.option}`, label: o.label }));

    const value =
      scope === 'public'
        ? getFieldValue(publicData, key)
        : scope === 'metadata'
        ? getFieldValue(metadata, key)
        : null;

    return isTargetEntityType && schemaType === SCHEMA_TYPE_MULTI_ENUM && displayInProfile
      ? [
          ...pickedElements,
          {
            schemaType,
            key,
            heading: config?.showConfig?.label,
            options: createFilterOptions(enumOptions),
            selectedOptions: value || [],
          },
        ]
      : isTargetEntityType && !!value && config.schemaType === SCHEMA_TYPE_TEXT && displayInProfile
      ? [
          ...pickedElements,
          {
            schemaType,
            key,
            heading: config?.showConfig?.label,
            text: value,
          },
        ]
      : pickedElements;
  }, []);
};
