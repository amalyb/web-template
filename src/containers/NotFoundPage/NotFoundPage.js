import React, { Component } from 'react';
import { arrayOf, bool, func, object, shape, string } from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { useConfiguration } from '../../context/configurationContext';
import { useRouteConfiguration } from '../../context/routeConfigurationContext';
import { FormattedMessage, useIntl, intlShape } from '../../util/reactIntl';
import { propTypes } from '../../util/types';
import { createResourceLocatorString } from '../../util/routes';
import { isScrollingDisabled } from '../../ducks/UI.duck';

import {
  Page,
  LayoutSingleColumn,
  LayoutWrapperTopbar,
  LayoutWrapperMain,
  LayoutWrapperFooter,
  Footer,
} from '../../components';
import TopbarContainer from '../../containers/TopbarContainer/TopbarContainer';

import LocationSearchForm from './LocationSearchForm/LocationSearchForm';

import css from './NotFoundPage.module.css';

export class NotFoundPageComponent extends Component {
  constructor(props) {
    super(props);
    // The StaticRouter component used in server side rendering
    // provides the context object. We attach a `notfound` flag to
    // the context to tell the server to change the response status
    // code into a 404.
    this.props.staticContext.notfound = true;
  }

  render() {
    const { history, routeConfiguration, siteTitle, intl, scrollingDisabled } = this.props;

    const title = intl.formatMessage({
      id: 'NotFoundPage.title',
    });

    const handleSearchSubmit = values => {
      const { search, selectedPlace } = values.location;
      const { origin, bounds } = selectedPlace;
      const searchParams = { address: search, origin, bounds };
      history.push(createResourceLocatorString('SearchPage', routeConfiguration, {}, searchParams));
    };

    return (
      <Page title={title} scrollingDisabled={scrollingDisabled}>
        <LayoutSingleColumn>
          <LayoutWrapperTopbar>
            <TopbarContainer />
          </LayoutWrapperTopbar>
          <LayoutWrapperMain>
            <div className={css.root}>
              <div className={css.content}>
                <div className={css.number}>404</div>
                <h1 className={css.heading}>
                  <FormattedMessage id="NotFoundPage.heading" />
                </h1>
                <p className={css.description}>
                  <FormattedMessage id="NotFoundPage.description" values={{ siteTitle }} />
                </p>
                <LocationSearchForm className={css.searchForm} onSubmit={handleSearchSubmit} />
              </div>
            </div>
          </LayoutWrapperMain>
          <LayoutWrapperFooter>
            <Footer />
          </LayoutWrapperFooter>
        </LayoutSingleColumn>
      </Page>
    );
  }
}

NotFoundPageComponent.defaultProps = {
  staticContext: {},
};

NotFoundPageComponent.propTypes = {
  scrollingDisabled: bool.isRequired,
  siteTitle: string.isRequired,

  // context object from StaticRouter, injected by the withRouter wrapper
  staticContext: object,

  // from useIntl
  intl: intlShape.isRequired,

  // from useRouteConfiguration
  routeConfiguration: arrayOf(propTypes.route).isRequired,

  // from useHistory
  history: shape({
    push: func.isRequired,
  }).isRequired,
};

const EnhancedNotFoundPage = props => {
  const routeConfiguration = useRouteConfiguration();
  const config = useConfiguration();
  const history = useHistory();
  const intl = useIntl();

  return (
    <NotFoundPageComponent
      routeConfiguration={routeConfiguration}
      siteTitle={config.siteTitle}
      history={history}
      intl={intl}
      {...props}
    />
  );
};

const mapStateToProps = state => {
  return {
    scrollingDisabled: isScrollingDisabled(state),
  };
};

// Note: it is important that the withRouter HOC is **outside** the
// connect HOC, otherwise React Router won't rerender any Route
// components since connect implements a shouldComponentUpdate
// lifecycle hook.
//
// See: https://github.com/ReactTraining/react-router/issues/4671
const NotFoundPage = compose(connect(mapStateToProps))(EnhancedNotFoundPage);

export default NotFoundPage;
