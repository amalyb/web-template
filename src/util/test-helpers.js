import React from 'react';
import mapValues from 'lodash/mapValues';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import configureStore from '../store';
import { IntlProvider } from '../util/reactIntl';
import defaultConfig from '../config/defaultConfig';
import { ConfigurationProvider } from '../context/configurationContext';
import { RouteConfigurationProvider } from '../context/routeConfigurationContext';
import routeConfiguration from '../routing/routeConfiguration';

// In case you have translated the template and have new translations that
// are missing from the en translations file, the language for the tests can
// be changed here so that there are no missing translation keys in tests.
import messages from '../translations/en.json';

Enzyme.configure({ adapter: new Adapter() });

let undefined;
export const getDefaultConfiguration = () => {
  return {
    ...defaultConfig,
    currency: 'USD',
    facebookAppId: undefined,
    canonicalRootURL: 'http://localhost:3000',
    maps: {
      ...defaultConfig.maps,
      mapboxAccessToken: undefined,
      googleMapsAPIKey: undefined,
    },
    stripe: {
      ...defaultConfig.stripe,
      publishableKey: 'pk_test_',
    },
  };
};

export const getRouteConfiguration = () => {
  const layoutConfig = {
    searchPageVariant: 'map',
    listingPageVariant: 'full-image',
    listingImage: {
      aspectWidth: 400,
      aspectHeight: 400,
      variantPrefix: 'listing-card',
    },
  };
  return routeConfiguration(layoutConfig);
};

// Locale should not affect the tests. We ensure this by providing
// messages with the key as the value of each message.
const testMessages = mapValues(messages, (val, key) => key);

// Provide all the context for components that connect to the Redux
// store, i18n, router, etc.
export const TestProvider = props => {
  const store = configureStore();
  return (
    <ConfigurationProvider value={getDefaultConfiguration()}>
      <RouteConfigurationProvider value={getRouteConfiguration()}>
        <IntlProvider locale="en" messages={testMessages} textComponent="span">
          <BrowserRouter>
            <Provider store={store}>{props.children}</Provider>
          </BrowserRouter>
        </IntlProvider>
      </RouteConfigurationProvider>
    </ConfigurationProvider>
  );
};

// Use Enzyme's shallow rendering to render the given component to a
// JSON structure that can be used in snapshot tests. This doesn't
// render the children within the given component, only a
// representation of the child component and its props.
//
// Useful for snapshot testing components that contain shared
// components. With deep rendering, if the child component changes
// internally, the test for the given component would also fail. This
// avoids the problem by not rendering the full tree but only the
// relevant structure for the given component.
export const renderShallow = component => {
  return toJson(shallow(component));
};

// Fully render the given component to a JSON structure that can be
// used in snapshot tests.
export const renderDeep = component => {
  return toJson(mount(<TestProvider>{component}</TestProvider>), { mode: 'deep' });
};
