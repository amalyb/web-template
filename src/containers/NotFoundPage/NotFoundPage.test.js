import React from 'react';
import { renderShallow } from '../../util/test-helpers';
import { fakeIntl } from '../../util/test-data';
import { NotFoundPageComponent } from './NotFoundPage';

const noop = () => null;

const routeConfiguration = [
  {
    path: '/',
    name: 'LandingPage',
    component: props => <div />,
  },
  {
    path: '/about',
    name: 'AboutPage',
    component: props => <div />,
  },
];

describe('NotFoundPageComponent', () => {
  it('matches snapshot', () => {
    const tree = renderShallow(
      <NotFoundPageComponent
        params={{ displayName: 'my-shop' }}
        history={{ push: noop }}
        location={{ search: '' }}
        scrollingDisabled={false}
        authInProgress={false}
        currentUserHasListings={false}
        isAuthenticated={false}
        onLogout={noop}
        onManageDisableScrolling={noop}
        sendVerificationEmailInProgress={false}
        onResendVerificationEmail={noop}
        intl={fakeIntl}
        routeConfiguration={routeConfiguration}
        siteTitle="MarketplaceX"
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
