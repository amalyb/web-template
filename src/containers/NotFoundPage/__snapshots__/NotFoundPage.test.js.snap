// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFoundPageComponent matches snapshot 1`] = `
<Page
  scrollingDisabled={false}
  title="NotFoundPage.title"
>
  <LayoutSingleColumn
    className={null}
    footer={<injectIntl(Footer) />}
    rootClassName={null}
    topbar={<withRouter(Connect(TopbarContainerComponent)) />}
  >
    <div
      className="root"
    >
      <div
        className="content"
      >
        <div
          className="number"
        >
          404
        </div>
        <Heading
          as="h1"
          className={null}
          rootClassName="heading"
        >
          <MemoizedFormattedMessage
            id="NotFoundPage.heading"
          />
        </Heading>
        <p
          className="description"
        >
          <MemoizedFormattedMessage
            id="NotFoundPage.description"
            values={
              Object {
                "marketplaceName": "MarketplaceX",
              }
            }
          />
        </p>
        <injectIntl(LocationSearchFormComponent)
          className="searchForm"
          isKeywordSearch={false}
          onSubmit={[Function]}
        />
      </div>
    </div>
  </LayoutSingleColumn>
</Page>
`;
